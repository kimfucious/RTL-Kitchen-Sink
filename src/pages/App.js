import React, { useState } from "react";
import $ from "jquery";
import { useSelector, useDispatch } from "react-redux";
import { reverseUsername, getRandomUserData } from "../actions.js/userActions";
import { Spinner } from "../components/shared/Spinner";
import { UserDataModal } from "../components/UserDataModal";
import { getTitleCase } from "../helpers";

export const App = () => {
  const dispatch = useDispatch();
  const {
    user,
    user: { username }
  } = useSelector((state) => state);

  const [animation, setAnimation] = useState(false);
  const [error, setError] = useState("");
  const [isSpinning, setIsSpinning] = useState(false);

  const handleReverse = async () => {
    try {
      setError("");
      setIsSpinning("reversing");
      setAnimation("");
      await dispatch(
        reverseUsername(user.username, setIsSpinning, setAnimation)
      );
      setIsSpinning(false);
      setAnimation("animate__animated animate__tada");
    } catch (error) {
      setIsSpinning(false);
      setError(error.message);
    }
  };

  const handleGetUserData = async () => {
    try {
      setError("");
      setIsSpinning("fetching");
      await dispatch(getRandomUserData());
      setIsSpinning(false);
      $("#userDataModal").modal("show");
    } catch (error) {
      setIsSpinning(false);
      setError(error.message);
    }
  };

  return (
    <>
      <UserDataModal />
      <div
        className="d-flex flex-column align-items-center justify-content-center w-100 full-height-with-nav animate__animated animate__fadeIn"
        data-testid="app-page"
      >
        <span className="display-1" role="img" aria-label="desert island">
          üèùÔ∏è
        </span>
        <div
          className={`d-flex justify-content-center display-3 text-center text-muted mt-3`}
        >
          <span>Aloha, </span>
          <span className={`d-flex ml-3 ${animation}`} data-testid="username">
            {getTitleCase(username)}
          </span>
          !
        </div>
        <div className="d-flex align-items-center justify-content-center mt-5">
          <button
            className="d-flex align-items-center justify-content-center btn btn-outline-primary mr-2"
            disabled={isSpinning === "reversing"}
            onClick={() => handleReverse()}
            style={{ height: 38, width: 128 }}
          >
            {isSpinning === "reversing" ? (
              <div>
                <Spinner size={10} color="primary" type="beat" />
              </div>
            ) : (
              <span>Reverse Name</span>
            )}
          </button>
          <button
            className="d-flex align-items-center justify-content-center btn btn-outline-secondary ml-2"
            disabled={isSpinning === "fetching"}
            onClick={() => handleGetUserData()}
            style={{ height: 38, width: 128 }}
          >
            {isSpinning === "fetching" ? (
              <div>
                <Spinner size={10} color="secondary" type="beat" />
              </div>
            ) : (
              <span>Get New User</span>
            )}
          </button>
        </div>
        <div
          className={`mt-3 animate__animated animate__${
            error ? "fadeIn" : "fadeOut"
          }`}
        >
          <p className="text-warning">{error}</p>
        </div>
      </div>
    </>
  );
};
