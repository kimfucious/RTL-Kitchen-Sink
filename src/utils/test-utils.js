import React from "react";
import { AppRouter } from "../routers/AppRouter";
import { Provider } from "react-redux";
import { render as rtlRender } from "@testing-library/react";
import configureStore from "../store/configureStore";

const render = (
  ui,
  {
    initialState = { auth: { userId: "", user: { username: "" } } },
    store = configureStore(initialState),
    ...renderOptions
  } = {}
) => {
  const Wrapper = ({ children }) => {
    return (
      <Provider store={store}>
        <AppRouter>{children}</AppRouter>
      </Provider>
    );
  };
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};

export * from "@testing-library/react";

export { render };
