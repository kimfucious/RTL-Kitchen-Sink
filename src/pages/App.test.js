import React from "react";
import { cleanup, render } from "../utils/test-utils";
import { App } from "./App";

const stateGeorge = {
  auth: { userId: "1234567890" },
  user: { username: "george" }
};
afterEach(cleanup);

describe("App", () => {
  test("renders", () => {
    const { asFragment } = render(<App />, { initialState: stateGeorge });
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders SignIn when userId is blank", () => {
    const { getByText } = render(<App />);
    const signInButton = getByText(/Sign In/);
    expect(signInButton).toBeInTheDocument();
  });

  test("renders App when userId is present", () => {
    const { getByText } = render(<App />, { initialState: stateGeorge });
    const reverseButton = getByText(/Reverse Name/);
    expect(reverseButton).toBeInTheDocument();
  });

  test("renders captitalized username", () => {
    const { getByText } = render(<App />, { initialState: stateGeorge });
    const username = getByText(/George/);
    expect(username).toBeInTheDocument();
  });
});
