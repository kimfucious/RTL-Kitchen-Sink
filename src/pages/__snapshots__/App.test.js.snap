// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-lg navbar-dark animate__animated animate__fadeIn"
  >
    <div
      class="container"
    >
      <a
        class="d-flex align-items-center navbar-brand"
        href="#"
      >
        <span
          aria-label="Octopus"
          class="mr-2"
          role="img"
          style="font-size: 32px;"
        >
          🐙
        </span>
        +
        <img
          alt="brand"
          class="ml-3"
          height="32"
          src="sink_32.png"
          width="32"
        />
      </a>
      <button
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-target="#navbarSupportedContent"
        data-toggle="collapse"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <ul
          class="navbar-nav mr-auto"
        />
        <button
          class="btn btn-outline-primary my-2 my-sm-0"
        >
          Sign Out
        </button>
      </div>
    </div>
  </nav>
  <div
    class="d-flex flex-column align-items-center justify-content-center w-100 full-height-with-nav animate__animated animate__fadeIn"
  >
    <div
      class="d-flex justify-content-center display-3 text-center text-muted"
    >
      <span
        aria-label="desert island"
        role="img"
      >
        🏝️
      </span>
      <span>
        Aloha, 
      </span>
      <span
        class="d-flex ml-3 false"
        data-testid="username"
      >
        George
      </span>
      !
    </div>
    <button
      class="d-flex align-items-center justify-content-center btn btn-outline-primary mt-4"
      style="height: 38px; width: 128px;"
    >
      <span>
        Reverse Name
      </span>
    </button>
    <div
      class="mt-3 animate__animated animate__fadeOut"
    >
      <p
        class="text-warning"
      />
    </div>
  </div>
</DocumentFragment>
`;
